server/index.ts

```diff
@@ -92,34 +120,51 @@ app.post("/default/davaiServer/message", async (req, res) => {
       }
     );
 
-    // Clean up the controller
-    activeMessages.delete(messageId);
-
     const lastMessage = output.messages[output.messages.length - 1];
     const response = await buildResponse(lastMessage);
+
+    // If this is a final response (no more tool calls needed), clean up the conversation
+    if (!("status" in response) || response.status !== "requires_action") {
+      activeConversations.delete(messageId);
+    }
+
     res.json(response);
   } catch (err: any) {
     if (err.message === "Aborted") {
-      res.status(500).json({ error: "Message processing cancelled" });
+      // Successful cancellation - return 200, not 500
+      res.status(200).json({
+        status: "cancelled",
+        message: "Message processing was successfully cancelled",
+        messageId
+      });
     } else {
       console.error("Error in /api/message:", err);
       console.error("Error stack:", err.stack);
       res.status(500).json({ error: "LangChain Error", details: err.message });
     }
-    activeMessages.delete(messageId);
+    activeConversations.delete(messageId);
   }
 });
 
 // Endpoint for cancelling message processing
 app.post("/default/davaiServer/cancel", async (req, res) => {
   const { messageId } = req.body;
-  const controller = activeMessages.get(messageId);
+  const controller = activeConversations.get(messageId);
   if (controller) {
     controller.abort();
-    activeMessages.delete(messageId);
-    res.json({ status: "cancelled", message: "Message processing cancelled successfully" });
+    activeConversations.delete(messageId);
+    res.json({
+      status: "cancelled",
+      messageId,
+      message: "Message processing cancelled successfully"
+    });
   } else {
-    res.status(404).json({ error: "Message not found or already completed" });
+    // Message not found could mean it already completed or never existed
+    res.status(404).json({
+      error: "Message not found or already completed",
+      messageId,
+      reason: "completed" // Message likely completed before cancellation request
```

Comment:
This is probably fine as is but I can't help but wonder if we should use "not found" instead of "completed" for the `reason` value? We don't know for sure it was completed. And the `error` value explains that even though the message wasn't found, it possibly already completed.